### PR TITLE
Python2 support for Scurl

### PR DESCRIPTION
**Pull request by nctl144**
This PR aims to support python2 on Scurl

### CODE REVIEW

**Comment by nctl144** [comment]
hey @lopuhin , I made this PR to support Py2 for this project. Let me know what you think :)

File: scurl/cgurl.pyx [hunk 1]
```diff
@@ -153,8 +153,13 @@ cdef object extra_attr(obj, prop, bytes url, Parsed parsed, decoded, params=Fals
...
     elif prop == "hostname":
         hostname = slice_component(url, parsed.host).lower()
-        if len(hostname) > 0 and chr(hostname[0]) == '[':
-            hostname = hostname[1:-1]
+        if len(hostname) > 0:
+            if six.PY2:
+                if hostname[0] == '[':
+                    hostname = hostname[1:-1]
+            else:
+                if chr(hostname[0]) == '[':
```
[the remark this answers is not in the document]
Comment (nctl144) [reply scurl/cgurl.pyx]: yep that works, thank you so much for the suggestion! 😄

File: scurl/cgurl.pyx [hunk 2]
```diff
@@ -329,14 +334,14 @@ class ParsedResultNamedTuple(tuple):
         if canonicalize and canonicalize_encoding != 'utf-8':
             if query:
                 try:
-                    query = query.decode().encode(canonicalize_encoding)
+                    query = query.decode('utf-8').encode(canonicalize_encoding)
                 except UnicodeEncodeError as e:
-                    pass
+                    print('Failed to encode query to the selected encoding!')
```
[the remark this answers is not in the document]
Comment (nctl144) [reply scurl/cgurl.pyx]: That makes sense! I'm fixing it now :)

File: tests/test_urlparse_py2.py [hunk 3]
```diff
@@ -202,22 +206,23 @@ def test_http_roundtrips(self):
...
                          (base, relurl, expected))
 
     def test_unparse_parse(self):
         for u in ['Python', './Python','x-newscheme://foo.com/stuff','x://y','x:/y','x:/','/',]:
-            self.assertEqual(urlparse.urlunsplit(urlparse.urlsplit(u)), u)
-            self.assertEqual(urlparse.urlunparse(urlparse.urlparse(u)), u)
+            self.assertEqual(scurl.urlunsplit(scurl.urlsplit(u)), u)
+            self.assertEqual(scurl.urlunparse(scurl.urlparse(u)), u)
 
+    @pytest.mark.xfail(reason='GURL failed to join http://a/b/c/d;p?q#f and " "')
```
[the remark this answers is not in the document]
Comment (nctl144) [reply tests/test_urlparse_py2.py]: Hey @lopuhin , it only failed the test listed in the reason, so I will move that test to another method now :)

File: tests/test_urlparse_py2.py [hunk 4]
```diff
@@ -396,6 +405,7 @@ def test_urljoins(self):
         self.checkJoin('svn://pathtorepo/dir1','dir2','svn://pathtorepo/dir2')
         self.checkJoin('svn+ssh://pathtorepo/dir1','dir2','svn+ssh://pathtorepo/dir2')
 
+    @pytest.mark.xfail(reason='marked as failed for now, it does not raise exception for invalid urls')
```
[the remark this answers is not in the document]
Comment (nctl144) [reply tests/test_urlparse_py2.py]: Yes it failed to raise exception for the invalid ipv6 urls such as http://::12.34.56.78]/. This is definitely what we will need to fix in the future. For now, I have moved it to another module at the end and mark the test that it belongs to :)

**Comment by nctl144** [comment]
Thank you, Konstantin!

**Comment by lopuhin** [comment]
Thank you @nctl144 ! Btw, I noticed that in setup.py we don't list python 3 compatibility, only 2.7: https://github.com/nctl144/scurl/blob/9264a1ed234a63ff720d6c9e68680fb3c7f044ca/setup.py#L84

**Comment by nctl144** [comment]
Yeah I haven't taken a look at the setup.py yet and it was still the one I got from urlparse4 😆 . I will fix it now :)

**Pull request merged**